loadbalancer/algorithm.go

```diff
@@ -143,17 +144,16 @@ func withFadeIn(rnd *rand.Rand, ctx *routing.LBContext, notFadingIndexes []int,
 }
 
 type roundRobin struct {
-	mx               sync.Mutex
-	index            int
+	index            int64
 	rnd              *rand.Rand
 	notFadingIndexes []int
 	fadingWeights    []float64
```

Comment:
Hm, looks like we have to guard notFadingIndexes and fadingWeights

Reply:
Yes, it makes sense to have mutex for every loadbalancer struct using `withFadeIn` function.
However, I'd like still to have `lockedSource` object not to forget to lock `rand.Source` in the future.

Reply:
@AlexanderYastrebov it looks like those fields are needed only for preallocation purposes, not to do actual writes to those fields.

Reply:
@AlexanderYastrebov withFadeIn writes to *copy* of it, like here:
```
rzavodskikh@ZALANDO-39125:/tmp$ cat copySlice.go 
package main

import (
	"fmt"
)

func foo(s []int) {
	s = append(s, 4)
}

func main() {
	a := []int{1, 2, 3}
	foo(a)

	fmt.Printf("a = %+v\n", a)
}

rzavodskikh@ZALANDO-39125:/tmp$ go build copySlice.go 
rzavodskikh@ZALANDO-39125:/tmp$ ./copySlice 
a = [1 2 3]
```

Reply:
https://go.dev/ref/spec#Appending_and_copying_slices

> If the capacity of s is not large enough to fit the additional values, append [allocates](https://go.dev/ref/spec#Allocation) a new, sufficiently large underlying array that fits both the existing slice elements and the additional values. Otherwise, append re-uses the underlying array.

Since field capacity is pre-allocated the code writes to the same array even though it uses append (which might be a bit confusing, I agree).

Reply:
Yeah, I do understand that append inside `withFadeIn` function may write (and probably writes) to the same **underlying array**, that's true.
However, since caller passes **slice** to the `withFadeIn` function, after the return this **slice** will remain the same (even though there were writes to the **underlying array**, you can search by "The output might be confusing!" here: https://trstringer.com/golang-slice-references/ )

Reply:
The only clue that I see is comment `// preallocating frequently used slice`, so it seems to me those slices were intended to improve performance of `Apply`.
However, when I tried to benchmark this operation, I see no significant difference with such fields and without them:
```
Every 10.0s: benchstat old.txt new.txt                                                                                                                                      ZALANDO-39125: Wed Jun 21 16:30:37 2023

goos: linux
goarch: amd64
pkg: github.com/zalando/skipper/loadbalancer
cpu: AMD Ryzen 7 PRO 4750U with Radeon Graphics
                                           │   old.txt   │              new.txt               │
                                           │   sec/op    │   sec/op     vs base               │
FadeIn/random,_10,_1_clients-16              830.2n ± 3%   871.6n ± 3%  +4.99% (p=0.000 n=74)
FadeIn/random,_10,_4_clients-16              1.147µ ± 8%   1.081µ ± 6%       ~ (p=0.084 n=74)
FadeIn/random,_10,_16_clients-16             1.261µ ± 4%   1.203µ ± 4%       ~ (p=0.169 n=74)
FadeIn/random,_10,_64_clients-16             1.044µ ± 4%   1.010µ ± 4%       ~ (p=0.084 n=74)
FadeIn/random,_10,_256_clients-16            1.071µ ± 3%   1.032µ ± 3%       ~ (p=0.139 n=74)
FadeIn/round-robin,_10,_1_clients-16         928.0n ± 3%   983.6n ± 3%  +6.00% (p=0.000 n=74)
FadeIn/round-robin,_10,_4_clients-16         762.3n ± 5%   778.2n ± 6%       ~ (p=0.527 n=74)
FadeIn/round-robin,_10,_16_clients-16        813.8n ± 6%   787.0n ± 4%       ~ (p=0.371 n=74)
FadeIn/round-robin,_10,_64_clients-16        633.2n ± 3%   618.3n ± 1%  -2.36% (p=0.043 n=74)
FadeIn/round-robin,_10,_256_clients-16       669.3n ± 5%   653.5n ± 2%       ~ (p=0.234 n=74)
FadeIn/consistent-hash,_10,_1_clients-16     1.169µ ± 3%   1.258µ ± 2%  +7.61% (p=0.000 n=74)
FadeIn/consistent-hash,_10,_4_clients-16     611.6n ± 2%   649.4n ± 4%  +6.17% (p=0.006 n=74)
FadeIn/consistent-hash,_10,_16_clients-16    525.1n ± 5%   545.9n ± 3%       ~ (p=0.234 n=74)
FadeIn/consistent-hash,_10,_64_clients-16    430.3n ± 2%   436.9n ± 3%       ~ (p=0.456 n=74)
FadeIn/consistent-hash,_10,_256_clients-16   448.9n ± 2%   450.3n ± 1%       ~ (p=0.533 n=74)
geomean                                      779.3n        782.4n       +0.40%
```